### PR TITLE
Remove duplicate identifiers in duplicate imports

### LINKED ISSUE #2437: Remove duplicates names in `no-duplicates` autofix
Let’s say we have the following content:

```js
import { a, b } from 'module'
import { a, c } from 'module'
```

The `import/no-duplicates` rule will report this, then perform the following autofix:

```js
import { a, b, a, c } from 'module'
```

I would expect it to remove duplicates, so the output would be:

```js
import { a, b, c } from 'module'
```

In my experience this is common when resolving merge conflicts if two branches added a new member to the import statement. It’s easy to _Accept both changes_, then let ESLint autofix resolve any issues with the imports. However, afterwards one has to remove the duplicates manually.

### PR DESCRIPTION
**Pull request by joe-matsec**
Resolves #2437.

This PR deduplicates import identifiers!

### CODE REVIEW

**Comment by joe-matsec** [comment]
Closing and resubmitting with my personal account 😅

**Comment by ljharb** [comment]
oof, duplicate PRs never go away :-/ can this be reopened so i can keep the two PRs in sync?

**Comment by joe-matsec** [comment]
Sorry about that.
Not sure I understand the idea behind keeping them both open but sure I can do that.

File: tests/src/rules/no-duplicates.js [hunk 1]
```diff
@@ -129,6 +129,20 @@ ruleTester.run('no-duplicates', rule, {
...
+      errors: ['\'./foo\' imported multiple times.', '\'./foo\' imported multiple times.'],
+    }),
+
+    // #2347: duplicate identifiers should be removed
+    test({
+      code: "import {a,b} from './foo'; import { b, c } from './foo'; import {b,c,d} from './foo'",
+      output: "import {a,b, c,d} from './foo';  ",
+      errors: ['\'./foo\' imported multiple times.', '\'./foo\' imported multiple times.', '\'./foo\' imported multiple times.'],
+    }),
+
```
Comment (joe-matsec) [inline tests/src/rules/no-duplicates.js:145]: I did build the plugin with my changes incorporated and tested it out manually to verify it works.

These tests should pass but they are failing due to parsing errors:

```
  2236 passing (7s)
  2 failing

  1) no-duplicates invalid import {a} from './foo'; import { a } from './foo':

      AssertionError [ERR_ASSERTION] [ERR_ASSERTION]: A fatal parsing error occurred: Parsing error: Identifier 'a' has already been declared
      + expected - actual

      -false
      +true
      
      at runRuleForItem (node_modules/eslint/lib/rule-tester/rule-tester.js:693:13)
      at testInvalidTemplate (node_modules/eslint/lib/rule-tester/rule-tester.js:800:28)
      at Context.<anonymous> (node_modules/eslint/lib/rule-tester/rule-tester.js:1039:29)
      at processImmediate (node:internal/timers:466:21)

  2) no-duplicates invalid import {a,b} from './foo'; import { b, c } from './foo'; import {b,c,d} from './foo':

      AssertionError [ERR_ASSERTION] [ERR_ASSERTION]: A fatal parsing error occurred: Parsing error: Identifier 'b' has already been declared
      + expected - actual

      -false
      +true
      
      at runRuleForItem (node_modules/eslint/lib/rule-tester/rule-tester.js:693:13)
      at testInvalidTemplate (node_modules/eslint/lib/rule-tester/rule-tester.js:800:28)
      at Context.<anonymous> (node_modules/eslint/lib/rule-tester/rule-tester.js:1039:29)
      at processImmediate (node:internal/timers:466:21)
```

I'm not sure how to proceed here?

**Comment by ljharb** [comment]
If you want, I can rewrite the commit so it comes from your personal account :-)

(github keeps a "ref" for every PR that's ever been opened, that isn't fetched by default, although i fetch them all. as such, closing a PR leaves an orphaned ref, and it makes my git log harder to understand)

File: src/rules/no-duplicates.js [hunk 2]
```diff
@@ -105,17 +104,32 @@ function getFix(first, rest, sourceCode) {
       closeBrace != null &&
       isPunctuator(sourceCode.getTokenBefore(closeBrace), ',');
     const firstIsEmpty = !hasSpecifiers(first);
+    const firstExistingIdentifiers = firstIsEmpty
+      ? new Set()
+      : sourceCode.text.slice(openBrace.range[1], closeBrace.range[0])
+        .split(',')
+        .reduce((acc, cur) => acc.add(cur.trim()), new Set());
```
Comment (ljharb) [inline src/rules/no-duplicates.js:111]:
```suggestion
      : new Set(sourceCode.text.slice(openBrace.range[1], closeBrace.range[0])
        .split(',')
        .map((x) => x.trim())
      );
```

File: src/rules/no-duplicates.js [hunk 3]
```diff
@@ -105,17 +104,32 @@ function getFix(first, rest, sourceCode) {
...
 
     const [specifiersText] = specifiers.reduce(
-      ([result, needsComma], specifier) => {
+      ([result, needsComma, existingIdentifiers], specifier) => {
+        // Add *only* the new identifiers that don't already exist, and track any new identifiers so we don't add them again in the next loop
+        const [specifierText, updatedExistingIdentifiers] = specifier.identifiers.reduce(([text, set], cur) => {
+          const trimmed = cur.trim(); // Trim whitespace before/after to compare to our set of existing identifiers
+          if (existingIdentifiers.has(cur.trim())) {
```
Comment (ljharb) [inline src/rules/no-duplicates.js:118]:
```suggestion
          if (existingIdentifiers.has(trimmed)) {
```

File: tests/src/rules/no-duplicates.js (hunk 1 above)
Comment (ljharb) [reply tests/src/rules/no-duplicates.js:145]: Hmm, that's a really good question. If duplicate identifiers aren't parseable at all, then i'm not sure how the plugin would be capable of linting them.

You could play with different parsers with the `parser: ` option, and see if maybe one (or all) of the babel or TS parsers can handle it?

Comment (joe-matsec) [reply tests/src/rules/no-duplicates.js]: That worked, thanks!

File: tests/src/rules/no-duplicates.js [hunk 4]
```diff
@@ -129,6 +129,30 @@ ruleTester.run('no-duplicates', rule, {
...
+      parser: parsers.BABEL_OLD, // Needed to avoid "A fatal parsing error occurred: Parsing error: Identifier 'b' has already been declared"
+    }),
+
+    // #2347: duplicate identifiers should be removed, but not if they are adjacent to comments
+    test({
+      code: "import {a} from './foo'; import { a/*,b*/ } from './foo'",
+      output: "import {a, a/*,b*/ } from './foo'; ",
+      errors: ['\'./foo\' imported multiple times.', '\'./foo\' imported multiple times.'],
+      parser: parsers.BABEL_OLD, // Needed to avoid "A fatal parsing error occurred: Parsing error: Identifier 'a' has already been declared"
+    }),
```
Comment (joe-matsec) [inline tests/src/rules/no-duplicates.js:154]: I added one more test case.
This feels a bit icky to me.
I could:

(A) leave it as-is
(B) treat this case as a "problematic comment" and bail (don't attempt to auto-fix)
(C) change the split logic for identifiers so it doesn't blindly split on commas and it takes comments into account -- this could get pretty complex, I guess I'd need a regex

What do you think?

Comment (ljharb) [reply tests/src/rules/no-duplicates.js:154]: It is almost always the proper choice to refuse to autofix any nodes adjacent to comments, because "comment attachment" (whether it points to the previous or the next node) is not a solved problem.

File: tests/src/rules/no-duplicates.js [hunk 5]
```diff
@@ -129,6 +129,30 @@ ruleTester.run('no-duplicates', rule, {
       errors: ['\'./foo\' imported multiple times.', '\'./foo\' imported multiple times.'],
     }),
 
+    // #2347: duplicate identifiers should be removed
+    test({
+      code: "import {a} from './foo'; import { a } from './foo'",
+      output: "import {a} from './foo'; ",
+      errors: ['\'./foo\' imported multiple times.', '\'./foo\' imported multiple times.'],
+      parser: parsers.BABEL_OLD, // Needed to avoid "A fatal parsing error occurred: Parsing error: Identifier 'a' has already been declared"
```
Comment (ljharb) [inline tests/src/rules/no-duplicates.js:137]: for all of these "babel old" test cases, do they *also* work on babel new, or on either of the two TS parsers?

if so, let's duplicate the test cases and cover each parser it works on :-)

File: tests/src/rules/no-duplicates.js (hunk 4 above)
Comment (joe-matsec) [reply tests/src/rules/no-duplicates.js:154]: Okay, so are you saying (A) the current behavior is acceptable?

What I meant by (B) was to change the logic inside `hasProblematicComments`:

```diff
diff --git a/src/rules/no-duplicates.js b/src/rules/no-duplicates.js
index bd094cb..c84a314 100644
--- a/src/rules/no-duplicates.js
+++ b/src/rules/no-duplicates.js
@@ -223,7 +223,7 @@ function hasProblematicComments(node, sourceCode) {
   return (
     hasCommentBefore(node, sourceCode) ||
     hasCommentAfter(node, sourceCode) ||
-    hasCommentInsideNonSpecifiers(node, sourceCode)
+    hasCommentInside(node, sourceCode)
   );
 }
 
@@ -241,18 +241,10 @@ function hasCommentAfter(node, sourceCode) {
     .some(comment => comment.loc.start.line === node.loc.end.line);
 }
 
-// Checks whether `node` has any comments _inside,_ except inside the `{...}`
-// part (if any).
-function hasCommentInsideNonSpecifiers(node, sourceCode) {
+// Checks whether `node` has any comments _inside_.
+function hasCommentInside(node, sourceCode) {
   const tokens = sourceCode.getTokens(node);
-  const openBraceIndex = tokens.findIndex(token => isPunctuator(token, '{'));
-  const closeBraceIndex = tokens.findIndex(token => isPunctuator(token, '}'));
-  // Slice away the first token, since we're no looking for comments _before_
-  // `node` (only inside). If there's a `{...}` part, look for comments before
-  // the `{`, but not before the `}` (hence the `+1`s).
-  const someTokens = openBraceIndex >= 0 && closeBraceIndex >= 0
-    ? tokens.slice(1, openBraceIndex + 1).concat(tokens.slice(closeBraceIndex + 1))
-    : tokens.slice(1);
+  const someTokens = tokens.slice(1);
   return someTokens.some(token => sourceCode.getCommentsBefore(token).length > 0);
 }
 ```

If you'd prefer I do this, I'm happy to do it, but three existing tests are failing and I'd need to change those too.

File: tests/src/rules/no-duplicates.js (hunk 5 above)
Comment (joe-matsec) [reply tests/src/rules/no-duplicates.js]: Done in d9ca6a238c283e23a614fce33a2e7aa327ff0749!

**Review by ljharb — CHANGES_REQUESTED** [review]
@joe-matsec sorry for the late review! I rebased this, and a bunch of tests are failing. It's entirely possible that my attempt to conflict resolve has caused the problem - can you take a look?

We may also need to add some more test cases for duplicate inline `type` imports.

**Comment by joe-matsec** [comment]
@ljharb I think this is ready for another review when you have some time!
I'm not sure how to deal with `Travis CI - Pull Request Action required after 4m — Build Errored` but everything else is green in CI.

**Comment by joe-matsec** [comment]
@ljharb I rebased again and it looks like CI is still good, if you have some time to review. Thanks!

**Review by ljharb — APPROVED** [review]
